File: ANSv4/Catalogo3/classes/EnvioDevolucion.vb
Option Explicit On

Public Class EnvioDevolucion

    ' Define como se llama este modulo para el control de errores
    Private Const m_sMODULENAME_ As String = "clsEnvioDevolucion"



    ' Esta clase realiza una consulta al servidor web y obtiene
    ' el numero de ip del servidor privado.
    Private Cliente As WSDevolucion.DevolucionSoapClient
    Private WebServiceInicializado As Boolean
    Private m_MacAddress As String
    Private m_ip As String

    Private m_NroDevolucion As String
    Private m_CodCliente As String
    Private m_Fecha As String
    Private m_Observaciones As String
    Private m_Detalle As String

    Private DatosObtenidos As Boolean

    Public ReadOnly Property Inicializado() As Boolean
        Get
            Return WebServiceInicializado
        End Get
    End Property

    '    Public Function ObtenerDatos(ByVal NroDevolucion As String)

    '        DatosObtenidos = False

    '        Dim rs_Enc As ADODB.Recordset
    '        Dim rs_Det As ADODB.Recordset

    '        rs_Enc = New ADODB.Recordset
    '        rs_Det = New ADODB.Recordset

    '        rs_Enc = adoModulo.adoComando(vg.Conexion, "EXECUTE v_Devolucion_Enc '" & NroDevolucion & "'")
    '        rs_Det = adoModulo.adoComando(vg.Conexion, "EXECUTE v_Devolucion_Det '" & NroDevolucion & "'")

    '        m_NroDevolucion = NroDevolucion
    '        m_CodCliente = Format(Trim(rs_Enc("IDCliente")), "000000")
    '        m_Fecha = rs_Enc("F_Devolucion")
    '        m_Observaciones = Replace(rs_Enc("Observaciones"), ",", " ") & ""

    '        'rs_Det.MoveFirst
    '        m_Detalle = ""
    '        Do Until rs_Det.EOF
    '0:          m_Detalle = m_Detalle & rs_Det("C_Producto") & ","                                '-00-
    '1:          m_Detalle = m_Detalle & Format(Trim(rs_Det("Cantidad")), "00000000") & "00,"      '-01-
    '2:          m_Detalle = m_Detalle & "NO" & ","                                                '-02-
    '3:          m_Detalle = m_Detalle & "NO" & ","                                                '-03-
    '4:          m_Detalle = m_Detalle & "NO" & ","                                                '-04-
    '            'm_Detalle = m_Detalle & IIf(Len(Trim(rs_Det("miDeposito"))) = 0, "   ", rs_Det("miDeposito")) & ","                                '-05-
    '5:          m_Detalle = m_Detalle & rs_Det("miDeposito") & ","
    '6:          m_Detalle = m_Detalle & rs_Det("Factura") & ","                                   '-06-
    '7:          m_Detalle = m_Detalle & Format(rs_Det("TipoDev"), "00") & ","                     '-07-
    '8:          m_Detalle = m_Detalle & rs_Det("Vehiculo") & ","                                  '-08-
    '9:          m_Detalle = m_Detalle & rs_Det("Modelo") & ","                                    '-09-
    '10:         m_Detalle = m_Detalle & rs_Det("Motor") & ","                                     '-10-
    '11:         m_Detalle = m_Detalle & rs_Det("Km") & ","                                        '-11-
    '12:         m_Detalle = m_Detalle & rs_Det("Observaciones") & ";"                             '-12-
    '            rs_Det.MoveNext()
    '        Loop

    '        rs_Enc = Nothing
    '        rs_Det = Nothing

    '        DatosObtenidos = True

    '    End Function

    Public Function EnviarDevolucion() As Long

        On Error GoTo ErrorHandler

        Dim Cancel As Boolean

        Cancel = False

        Dim resultado As Long

        If Not WebServiceInicializado Then
            Cancel = True
        End If

        If Not Cancel Then

            If Not (vg.TranActiva Is Nothing) Then
                vg.TranActiva.Rollback()
                vg.TranActiva = Nothing
            End If

            resultado = Cliente.EnviarDevolucion(m_MacAddress, m_NroDevolucion, m_CodCliente, m_Fecha, m_Observaciones, m_Detalle)

            If resultado = 0 Then
                '                adoModulo.adoComandoIU(vg.Conexion, "EXEC usp_Devolucion_Transmicion_Upd '" & m_NroDevolucion & "'")
                If Not (vg.TranActiva Is Nothing) Then
                    vg.TranActiva.Commit()
                    vg.TranActiva = Nothing
                End If
            Else
                If Not (vg.TranActiva Is Nothing) Then
                    vg.TranActiva.Rollback()
                    vg.TranActiva = Nothing
                End If
            End If

            EnviarDevolucion = resultado
        Else
            EnviarDevolucion = -1
        End If

        Exit Function

ErrorHandler:
        Err.Raise(Err.Number, Err.Source, Err.Description)

    End Function

    Public Sub Inicializar(ByVal ipAddress As String, _
                           ByVal ipAddressIntranet As String, _
                           ByVal MacAddress As String)

        Dim Conectado As Boolean
        Conectado = My.Computer.Network.Ping(ipAddress, 5000)
        Conectado = True
        If Not Conectado Then
            Conectado = My.Computer.Network.Ping(ipAddressIntranet, 5000)
        End If

        On Error GoTo ErrorHandler

        If Conectado Then
            If Not WebServiceInicializado Then
                Cliente = New WSDevolucion.DevolucionSoapClient("", "http://" & ipAddress & "/wsCatalogo3/Devolucion.asmx?wsdl")
                If Trim(Funciones.modINIs.ReadINI("datos", "proxy")) = "1" Then
                    '                Cliente.ConnectorProperty("ProxyServer") = Funciones.modINIs.ReadINI("datos", "proxyserver")
                End If

                m_MacAddress = MacAddress
                m_ip = ipAddress
                WebServiceInicializado = True
            End If
        Else
            WebServiceInicializado = False
        End If

        Exit Sub

ErrorHandler:

        If Err.Number = -2147024809 Then
            ' Intento con el ip interno
            Cliente = New WSDevolucion.DevolucionSoapClient("", "http://" & ipAddressIntranet & "/wsCatalogo3/Devolucion.asmx?wsdl")
            If Trim(Funciones.modINIs.ReadINI("datos", "proxy")) = "1" Then
                '                Cliente.ConnectorProperty("ProxyServer") = Funciones.modINIs.ReadINI("datos", "proxyserver")
            End If

            m_MacAddress = MacAddress
            m_ip = ipAddressIntranet
            WebServiceInicializado = True
            Err.Clear()
        Else
            Err.Raise(Err.Number, Err.Source, Err.Description)
        End If

    End Sub

End Class

File: ANSv4/Catalogo4/_devoluciones/EnvioDevolucion.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Collections;
using System.Data;
using System.Diagnostics;
//using Microsoft.VisualBasic;

namespace Catalogo._devoluciones
{
    public class EnvioDevolucion
    {
        // Esta clase realiza una consulta al servidor web y obtiene
        // el numero de ip del servidor privado.
        private System.Data.OleDb.OleDbTransaction _TranActiva = null;

        private DevolucionWS.Devolucion Cliente;
        private bool WebServiceInicializado;
        private string m_MacAddress;

        private string m_ip;
        private string m_NroDevolucion;
        private string m_CodCliente;
        private string m_Fecha;
        private string m_Observaciones;

        private string m_Detalle;


        public EnvioDevolucion(System.Data.OleDb.OleDbConnection conexion, string ipAddress, string MacAddress)
        {
            Inicializar(ipAddress, MacAddress);
            Conexion1 = conexion;
        }

        public bool Inicializado
        {
            get { return WebServiceInicializado; }
        }


        private System.Data.OleDb.OleDbConnection Conexion1;

        internal void ObtenerDatos(string NroDevolucion)
        {

            System.Data.OleDb.OleDbDataReader Enc = null;
            System.Data.OleDb.OleDbDataReader Det = null;

            Enc = Funciones.oleDbFunciones.Comando(Conexion1, "EXECUTE v_Devolucion_Enc '" + NroDevolucion + "'");
            Det = Funciones.oleDbFunciones.Comando(Conexion1, "EXECUTE v_Devolucion_Det '" + NroDevolucion + "'");

            m_NroDevolucion = NroDevolucion;
            Enc.Read();
            m_CodCliente = Enc["IDCliente"].ToString().Trim().PadLeft(6, '0');
            m_Fecha = string.Format("{0:yyyyMMdd}", DateTime.Parse(Enc["F_Devolucion"].ToString()));

            m_Observaciones = Enc["Observaciones"].ToString().Replace(",", " ");

            if (Det.HasRows)
            {
                m_Detalle = "";
                while (Det.Read())
                {
                    m_Detalle += Det["C_Producto"].ToString() + ",";
                    m_Detalle += Det["Cantidad"].ToString().Trim().PadLeft(8, '0') + "00,";
                    m_Detalle += "NO" + ",";
                    m_Detalle += "NO" + ",";
                    m_Detalle += "NO" + ",";
                    m_Detalle += Det["miDeposito"].ToString() + ",";
                    m_Detalle += Det["Factura"].ToString() + ",";
                    m_Detalle += Det["TipoDev"].ToString().Trim().PadLeft(2, '0') + ",";
                    m_Detalle += Det["Vehiculo"].ToString() + ",";
                    m_Detalle += Det["Modelo"].ToString() + ",";
                    m_Detalle += Det["Motor"].ToString() + ",";
                    m_Detalle += Det["Km"].ToString() + ",";
                    m_Detalle += Det["Observaciones"].ToString() + ";";
                }
            }

        }

        public long EnviarDevolucion()
        {
            bool Cancel = false;

            long resultado = 0;

            if (!WebServiceInicializado)
            {
                Cancel = true;
            }

            try
            {
                if (!Cancel)
                {
                    if (_TranActiva == null)
                    {
                        //@ _TranActiva =Conexion1.BeginTransaction();
                        util.errorHandling.ErrorLogger.LogMessage("2");
                    }

                    resultado = Cliente.EnviarDevolucion(m_MacAddress, m_NroDevolucion, m_CodCliente, m_Fecha, m_Observaciones, m_Detalle);

                    if (resultado == 0)
                    {
                        Funciones.oleDbFunciones.ComandoIU(Conexion1, "EXEC usp_Devolucion_Transmicion_Upd '" + m_NroDevolucion + "'");
                        if (_TranActiva != null)
                        {
                            _TranActiva.Commit();
                            _TranActiva = null;
                        }
                    }
                    else
                    {
                        if (_TranActiva != null)
                        {
                            _TranActiva.Rollback();
                            _TranActiva = null;
                            util.errorHandling.ErrorLogger.LogMessage("10 rollback");
                        }
                    }

                    return resultado;
                }
                else
                {
                    return -1;
                }
            }
            catch (Exception ex)
            {
                util.errorHandling.ErrorLogger.LogMessage(ex);
                return -1;
            }
        }


        internal void Inicializar(string ipAddress, string MacAddress)
        {
            bool Conectado = util.network.IPCache.instance.conectado;

            try
            {
                if (Conectado)
                {
                    if (!WebServiceInicializado)
                    {
                        Cliente = new DevolucionWS.Devolucion();
                        Cliente.Url = "http://" + ipAddress + "/wsCatalogo4/Devolucion.asmx?wsdl";
                        if (Global01.proxyServerAddress != "0.0.0.0")
                        {
                            Cliente.Proxy = new System.Net.WebProxy(Global01.proxyServerAddress);
                        }

                        m_MacAddress = MacAddress;
                        m_ip = ipAddress;
                        WebServiceInicializado = true;
                    }
                }
                else
                {
                    WebServiceInicializado = false;
                }
            }
            catch (Exception ex)
            {
                //if (System.Runtime.InteropServices.Marshal.GetExceptionCode() == -2147024809)
                //{
                //    //	if (Err().Number == -2147024809) {
                //    // Intento con el ip interno
                //    Cliente = new DevolucionWS.Devolucion();
                //    Cliente.Url = "http://" + ipAddressIntranet + "/wsCatalogo4/Devolucion.asmx?wsdl";
                //    if (Global01.proxyServerAddress != "0.0.0.0")
                //    {
                //        Cliente.Proxy = new System.Net.WebProxy(Global01.proxyServerAddress);
                //    }

                //    m_MacAddress = MacAddress;
                //    m_ip = ipAddressIntranet;
                //    WebServiceInicializado = true;
                //}
                //else
                //{
                throw ex;
                //}
            }
        }

    }
}